sourced/ml/transformers/basic.py

```diff
@@ -283,8 +293,13 @@ def create_uast_source(args, session_name, select=HeadFiles, language_selector=N
         parquet_loader_args = filter_kwargs(args.__dict__, create_parquet_loader)
         start_point = create_parquet_loader(session_name, **parquet_loader_args)
         root = start_point
-        if extract_uast and "uast" not in [col.name for col in start_point.execute().schema]:
+        fields = [col.name for col in start_point.execute().schema]
+        if extract_uast and "uast" not in fields:
             raise ValueError("The parquet files do not contain UASTs.")
+        if language_selector is not None:
+            start_point = start_point.link(language_selector)
```

Comment:
Why do we have to allow custom `LanguageSelector` in the first place?

The code becomes very complex and even contradictory. I would remove `language_selector` completely. Everything what it does can be repeated with `args.languages`. 

`select` arg is also breaking the abstraction because this method is command level and can be used as API level.

Reply:
I agree the code is really messy right now. 
We need to rewrite it somehow. 

We allow custom `LanguageSelector` because of only one case here: https://github.com/src-d/ml/blob/master/sourced/ml/cmd/repos2ids.py#L13 

Any suggestions what we can do?

Also `LanguageSelector` itself becomes complex. may be we can break it into `LanguageExtractor` and `LanguageFilter`.

Reply:
Then better to add `--blacklist` and remove the argument and the use case.